src/ffi/dictobject.rs

```diff
@@ -130,6 +64,60 @@ extern "C" {
     ) -> c_int;
     #[cfg_attr(PyPy, link_name = "PyPyDict_DelItemString")]
     pub fn PyDict_DelItemString(dp: *mut PyObject, key: *const c_char) -> c_int;
-    #[cfg(not(PyPy))]
-    pub fn _PyObject_GetDictPtr(obj: *mut PyObject) -> *mut *mut PyObject;
+// skipped 3.10 / ex-non-limited PyObject_GenericGetDict
+}
+
+#[cfg_attr(windows, link(name = "pythonXY"))]
+extern "C" {
+    pub static mut PyDictKeys_Type: PyTypeObject;
+    pub static mut PyDictValues_Type: PyTypeObject;
+    pub static mut PyDictItems_Type: PyTypeObject;
+}
+
+#[inline]
+pub unsafe fn PyDictKeys_Check(op: *mut PyObject) -> c_int {
+    (Py_TYPE(op) == &mut PyDictKeys_Type) as c_int
+}
+
+#[inline]
+pub unsafe fn PyDictValues_Check(op: *mut PyObject) -> c_int {
+    (Py_TYPE(op) == &mut PyDictValues_Type) as c_int
+}
+
+#[inline]
+pub unsafe fn PyDictItems_Check(op: *mut PyObject) -> c_int {
+    (Py_TYPE(op) == &mut PyDictItems_Type) as c_int
+}
+
+#[inline]
+pub unsafe fn PyDictViewSet_Check(op: *mut PyObject) -> c_int {
+    (PyDictKeys_Check(op) != 0 || PyDictItems_Check(op) != 0) as c_int
+}
+
+#[cfg_attr(windows, link(name = "pythonXY"))]
+extern "C" {
+    pub static mut PyDictIterKey_Type: PyTypeObject;
+    pub static mut PyDictIterValue_Type: PyTypeObject;
+    pub static mut PyDictIterItem_Type: PyTypeObject;
+    #[cfg(Py_3_8)]
+    pub static mut PyDictRevIterKey_Type: PyTypeObject;
+    #[cfg(Py_3_8)]
+    pub static mut PyDictRevIterValue_Type: PyTypeObject;
+    #[cfg(Py_3_8)]
+    pub static mut PyDictRevIterItem_Type: PyTypeObject;
 }
+
+#[repr(C)]
+#[derive(Debug)]
+// TODO: move to cpython/dictobject.rs
+// Not moved because dict.rs uses PyDictObject extensively.
+pub struct PyDictObject {
```

Comment:
@davidhewitt re "TODO"s:

`PyDictObject` is a (non-limited) CPython API, but used by dict.rs unconditionally. I've therefore left it in ffi/dictobject.rs, even though the definition is from CPython.

I'm not sure removing this dependency can fall within the scope of #1289. Do you have any suggestions as to how to deal with it? My inclination is to leave it in place for now and spawn another issue to address it if necessary.